// @babel/core/src/config/files/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
Promise.resolve().then(function () { return require("./index-browser"); });
Promise.resolve().then(function () { return require("./index"); });
// Kind of gross, but essentially asserting that the exports of this module are the same as the
// exports of index-browser, since this file may be replaced at bundle time with index-browser.
({});
var package_1 = require("./package");
exports.findPackageData = package_1.findPackageData;
var configuration_1 = require("./configuration");
exports.findConfigUpwards = configuration_1.findConfigUpwards;
exports.findRelativeConfig = configuration_1.findRelativeConfig;
exports.findRootConfig = configuration_1.findRootConfig;
exports.loadConfig = configuration_1.loadConfig;
exports.resolveShowConfigPath = configuration_1.resolveShowConfigPath;
exports.ROOT_CONFIG_FILENAMES = configuration_1.ROOT_CONFIG_FILENAMES;
type;
{
    ConfigFile,
        IgnoreFile,
        RelativeConfig,
        FilePackageData,
    ;
}
from;
"./types";
var plugins_1 = require("./plugins");
exports.loadPlugin = plugins_1.loadPlugin;
exports.loadPreset = plugins_1.loadPreset;
const gensync_1 = require("gensync");
const plugins = require("./plugins");
exports.resolvePlugin = gensync_1.default(plugins.resolvePlugin).sync;
exports.resolvePreset = gensync_1.default(plugins.resolvePreset).sync;
